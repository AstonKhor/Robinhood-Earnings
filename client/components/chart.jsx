import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import { Header, D3Create } from './index.js';

class Chart extends React.Component {
  constructor() {
    super()
  }

  componentDidMount() {
    fetch('/getData', {
      type: "GET",
    }).then((response) => {
      return response.json()
    }).then((response) => {
      console.log(response)
      this.setState({
        text: JSON.stringify(response)
      })
    })
  }
  render(){
    return(
      <React.Fragment>
        <Header />
        <D3Create />
      </React.Fragment>
    )
  }
}
export default Chart;
