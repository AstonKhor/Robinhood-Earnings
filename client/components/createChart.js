import * as d3 from 'd3';

const chart = (data) => {
  const height = 120;
  const width = 560;
  // const margin = ({top: 20, right: 30, bottom: 30, left: 40})
  const margin = ({top: 0, right: 0, bottom: 0, left: 0})
  const x = d3.scalePoint()
    .domain(data.map(d => d.x))
    .range([margin.left, width - margin.right])
  const y = d3.scaleLinear()
      .domain(d3.extent(data, d => d.y))
      .range([height - margin.bottom, margin.top])
  const xAxis = g => g
    .attr("transform", `translate(0,${height+ 45})`)
    .call(d3.axisBottom(x))
    .call(g => g.select(".domain").remove())
    .call(g => g.append("text")
        .attr("x", width - margin.right)
        .attr("y", -4)
        .attr("text-anchor", "end")
        .attr("font-size", 13)
        .attr("font-weight", 400)
        .attr("line-height", "19px")
        .text(data.name))
    .call(g => g.selectAll(".tick text")
        .attr("stroke", "#8c8c8e"))
    .call(g => g.selectAll(".tick line")
        .remove())

  const yAxis = g => g
      .attr("transform", `translate(${-51},0)`)
      .call(d3.axisLeft(y).ticks(4))
      .call(g => g.select(".domain").remove())
      .call(g => g.select(".tick:last-of-type text").clone()
          .attr("x", 4)
          .attr("text-anchor", "start")
          .attr("line-height", "19px")
          .attr("font-family", "arial")
          .attr("font-size", 13)
          .attr("font-weight", 400)
          .text(data.y))
      .call(g => g.selectAll(".tick text")
          .attr("stroke", "#8c8c8e"))
      .call(g => g.selectAll(".tick line")
          .remove())

  const svg = d3.select('#earnings').append('svg')
    .attr("width", width)
    .attr("height", height)
    .attr("overflow", "visible")

  svg.append("g")
    .call(xAxis)
  svg.append("g")
    .call(yAxis)
  svg.append("g")
      .attr("stroke-width", 1.5)
      .attr("font-family", "arial")
      .attr("font-size", 13)
      .attr("font-weight", 400)
    .selectAll("g")
    .data(data)
    .join("g")
      .attr("transform", d => `translate(${x(d.x)},${y(d.y)})`)
      .attr("opacity", d => {
        return (d.type === "estimated" ? "20%" : "")
      })
      .call((g) => g.append("circle")
          .attr('class', "")
          .attr("fill", "#f45531")
          .attr("r", 7))
  svg.selectAll('circle')
    .on("mouseover", handleMouseOver)
    .on("mouseout", handleMouseOut)
  function handleMouseOver(d, i) {
    d3.select(this).transition()
      .duration('100')
      .attr('r', '25');
    let len = data.length
    let otherNode;
    if (i < 8) {
      otherNode = i + 8
    } else {
      otherNode = i - 8
    }
    let diff = d.y - data[otherNode].y;
    // d3.select(this).append("text")
      // .attr({x: () => x(d.x)})
      // .attr({y: () => y(d.y)})
      // .text("asdf")
    svg.append("text")
      .attr('transform', `translate(${x(d.x)},${y(d.y) - 25})`)
      .attr('class', `ID${i}`)
      .attr('fill', 'white')
      .text(`${diff.toFixed(2)}`)
  }
  function handleMouseOut(d, i) {
    d3.select(this).transition()
      .duration('100')
      .attr('r', '7')
    d3.select(`.ID${i}`).remove()

  }
}
export default chart
