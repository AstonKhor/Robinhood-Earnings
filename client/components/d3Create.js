import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import * as d3 from 'd3';

const ChartArea = styled.div`
  padding: 10px 10px 74px 85px;
`
const chart = (data) => {
  const height = 120;
  const width = 560;
  // const margin = ({top: 20, right: 30, bottom: 30, left: 40})
  const margin = ({top: 0, right: 0, bottom: 0, left: 0})
  const x = d3.scalePoint()
    .domain(data.map(d => d.x))
    .range([margin.left, width - margin.right])
  const y = d3.scaleLinear()
      .domain(d3.extent(data, d => d.y))
      .range([height - margin.bottom, margin.top])
  const xAxis = g => g
    .attr("transform", `translate(0,${height+ 45})`)
    .call(d3.axisBottom(x))
    .call(g => g.select(".domain").remove())
    .call(g => g.append("text")
        .attr("x", width - margin.right)
        .attr("y", -4)
        .attr("text-anchor", "end")
        .attr("font-size", 13)
        .attr("font-weight", 400)
        .attr("line-height", "19px")
        .text(data.name))
    .call(g => g.selectAll(".tick text")
        .attr("stroke", "#8c8c8e"))
    .call(g => g.selectAll(".tick line")
        .remove())

  const yAxis = g => g
      .attr("transform", `translate(${-51},0)`)
      .call(d3.axisLeft(y))
      .call(g => g.select(".domain").remove())
      .call(g => g.select(".tick:last-of-type text").clone()
          .attr("x", 4)
          .attr("text-anchor", "start")
          .attr("line-height", "19px")
          .attr("font-family", "arial")
          .attr("font-size", 13)
          .attr("font-weight", 400)
          .text(data.y))
      .call(g => g.selectAll(".tick text")
          .attr("stroke", "#8c8c8e"))
      .call(g => g.selectAll(".tick line")
          .remove())

  const svg = d3.select('#earnings').append('svg')
    .attr("width", width)
    .attr("height", height)
    .attr("overflow", "visible")

  svg.append("g")
    .call(xAxis)
  svg.append("g")
    .call(yAxis)
  svg.append("g")
      .attr("stroke-width", 1.5)
      .attr("font-family", "arial")
      .attr("font-size", 13)
      .attr("font-weight", 400)
    .selectAll("g")
    .data(data)
    .join("g")
      .attr("transform", d => `translate(${x(d.x)},${y(d.y)})`)
      .attr("opacity", d => {
        return (d.type === "estimated" ? "20%" : "")
      })
      .call(g => g.append("circle")
          .attr("fill", "#f45531")
          .attr("r", 7))
}

const formatData = (data) => {
  let formatted = data.map((data, idx) => {
    return {
      name: data.date,
      x: data.date,
      y: data.estimated,
      type: "estimated"
    }
  })
  data.forEach((data, idx) => {
    formatted.push({
      name: data.date,
      x: data.date,
      y: data.actual,
      type: "actual"
    })
  })
  return formatted
}
const D3Create = (props) => {
  let { data } = props
  let formatted = formatData(data)
  chart(formatted)
  return (
    <ChartArea>
      <div id="earnings" />
    </ChartArea>
  )
}


export default D3Create;
