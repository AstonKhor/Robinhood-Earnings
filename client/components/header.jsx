import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';

const Title = styled.div`
  color: #fff;
  font-size: 26px;
  font-weight: 500;
  letter-spacing: -0.14px;
  line-height: 30px;
  border-bottom: 1px solid #0e0d0d;
  padding-bottom: 16px;
  margin-bottom: 24px;
`;


const Header = () => {
  return(
    <Title>Earnings</Title>
  )
}


export default Header;
