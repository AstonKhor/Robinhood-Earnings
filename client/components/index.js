import Chart from './chart';
import D3Create from './d3Create';
import Header from './header';

export { Chart, Header, D3Create }
