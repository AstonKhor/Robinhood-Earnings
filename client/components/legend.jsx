import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';

const Box = styled.div`
  align-items: center;
  display: flex;
  height: 64px;
  margin-top: 8px;
  color: #8c8c8e;
`;

const Estimated = styled.div`
  padding-left: 21px;
  margin-right: 44px;
  display: flex;
`

const Actual = styled.div`
  display: flex;
`

const Text = styled.div`
  margin-bottom: 0px;
  font-weight: 400;
`

const CircleContainer = styled.svg`
  margin-right: 12px;
  overflow: visible;
  color: #21ce99;
`
const Circle = styled.circle`
  fill: #21ce99;
`

const Header = styled.div`
  font-weight: 500;
`


const Legend = () => {
  return(
    <Box>
      <Estimated>
        <CircleContainer width="28" height="28">
          <g transform="translate(14,20)">
            <Circle r="7" opacity="0.2" />
          </g>
        </CircleContainer>
        <Text>
          <Header>
            Estimated
          </Header>
          <div>â€”</div>
        </Text>
      </Estimated>
      <Actual>
        <CircleContainer width="28" height="28">
          <g transform="translate(14,20)">
            <Circle r="7" />
          </g>
        </CircleContainer>
        <Text>
          <Header>
            Actual
          </Header>
          <div>Expected Jan 30, Pre-Market</div>
        </Text>
      </Actual>
    </Box>
  )
}


export default Legend;
