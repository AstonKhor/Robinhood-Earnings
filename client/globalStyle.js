import { createGlobalStyle } from 'styled-components';
import path from 'path';

const GlobalStyle = createGlobalStyle``

export default GlobalStyle
