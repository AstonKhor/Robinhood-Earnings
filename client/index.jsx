import React from 'react';
import ReactDOM from 'react-dom';
import GlobalStyle from './globalStyle.js';

import { Chart } from './components';

const App = () => {
  return(
    <React.Fragment>
      <GlobalStyle />
      <Chart />
    </React.Fragment>
  )
}

ReactDOM.render( <App /> , document.getElementById('earnings'))
