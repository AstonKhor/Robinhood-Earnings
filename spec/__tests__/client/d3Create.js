import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { D3Create } from '../../../client/components';
const data = [
	{
		"date" : "Q1 2018",
		"estimated" : 1.2338543412859448,
		"actual" : 1.2159874223729008
	},
	{
		"date" : "Q2 2018",
		"estimated" : 0.7628355868248811,
		"actual" : 0.8891278100932136
	},
	{
		"date" : "Q3 2018",
		"estimated" : 0.6082710612614539,
		"actual" : 1.256231485885062
	},
	{
		"date" : "Q4 2018",
		"estimated" : 0.33177004366859464,
		"actual" : 0.10239494600374677
	},
	{
		"date" : "Q1 2019",
		"estimated" : 0.6740910259085577,
		"actual" : 0.2845353248552196
	},
	{
		"date" : "Q2 2019",
		"estimated" : 0.47308762981075114,
		"actual" : 0.7641997896857413
	},
	{
		"date" : "Q3 2019",
		"estimated" : 1.6850565175210557,
		"actual" : 0.4614122638085081
	},
	{
		"date" : "Q4 2019",
		"estimated" : 1.2552326465002954,
		"actual" : 1.0327321625979917
	}
]
describe('d3 create method', () => {
  it('should render properly', () => {
    const component = renderer.create(<D3Create data={[data[0]]} />)
    expect(component).toMatchSnapshot();
  })

  it('should take in data as a prop', () => {
    const component = mount(<D3Create data={data} />)
    expect(component.props().data).toBeDefined();
    expect(component.props().data.length).toBeGreaterThan(1);
  })

  it('should render a chart with correct data', () => {
    // TODO not entirely sure what this will look like since the D3 isn't built yet.
  })

  it('should be green if trending up', () => {

  })

  it('should be red if trending down', () => {

  })

  it('should have only past 2 years of quarterly data', () => {

  })

  it('should have two keys of "Estimated" and "Actual"', () => {

  })
})
