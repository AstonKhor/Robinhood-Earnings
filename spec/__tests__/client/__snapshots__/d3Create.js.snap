// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`d3 create method should render properly 1`] = `
<div
  className="sc-bdVaJa ceTkBM"
>
  <div
    id="earnings"
  />
</div>
`;
